### PR TITLE
fetch sandbox (part of #498)

### PR DESCRIPTION
**Pull request by miherlosev**
@inikulin @churkin @VasilyStrelyaev

### CODE REVIEW

File: src/client/sandbox/fetch.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import nativeMethods from './native-methods';
+import XHR_HEADERS from '../../request-pipeline/xhr/headers';
+import { getProxyUrl, parseProxyUrl } from '../utils/url';
+import { getOrigin, sameOriginCheck, get as getDestLocation } from '../utils/destination-location';
+import { isHeaders, isRequest } from '../utils/dom';
+
+export default class FetchSandbox extends SandboxBase {
+    constructor () {
+        super();
+    }
```
Comment (churkin) [inline src/client/sandbox/fetch.js]: unnecessary

File: src/client/utils/dom.js [hunk 2]
```diff
@@ -542,6 +542,30 @@ export function isSVGElementOrChild (el) {
     return !!closest(el, 'svg');
 }
 
+export function isHeaders (instance) {
+    if (instance instanceof nativeMethods.Headers)
+        return true;
+
+    return instance && typeof instance === 'object' && instance.getAll !== void 0 &&
+           instance.toString() && instance.toString() === '[object Headers]';
```
Comment (churkin) [inline src/client/utils/dom.js]: `instance` can not have `toString` method

File: src/client/sandbox/fetch.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        init.headers = headers;
+
+        return init;
+    }
+
+    static _processArguments (args) {
+        var input = args[0];
+        var init  = args[1] || {};
```
Comment (churkin) [inline src/client/sandbox/fetch.js]: I propose to use the `options` term instead of `init`

File: src/client/sandbox/fetch.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import { getOrigin, sameOriginCheck, get as getDestLocation } from '../utils/destination-location';
+import { isHeaders, isRequest } from '../utils/dom';
+
+export default class FetchSandbox extends SandboxBase {
+    constructor () {
+        super();
+    }
+
+    static _processRequestInit (init) {
+        var requestDefaults    = new nativeMethods.Request('');
```
Comment (churkin) [inline src/client/sandbox/fetch.js]: `defaultRequest` &

**Comment by churkin** [comment]
-\r

File: src/client/sandbox/fetch.js (hunk 3 above)
Comment (miherlosev) [inline src/client/sandbox/fetch.js]: In specification `Request` parameter is named as `input` and `init` (https://fetch.spec.whatwg.org/#request-class).

File: test/client/fixtures/sandbox/fetch-test.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        .then(function (response) {
+            return response.json();
+        })
+        .then(function (json) {
+            strictEqual(JSON.stringify(json), JSON.stringify(data));
+            start();
+        });
+    });
+
+    asyncTest('internal 222 status code should be replaced to 0 on the client side', function () {
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/fetch-test.js]:
```
    asyncTest('the internal 222 status code should be replaced with 0 on the client side', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by miherlosev** [comment]
FPR

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 6]
```diff
@@ -575,14 +593,14 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
             },
 
             status: {
-                condition: XhrSandbox.isOpenedXhr,
+                condition: instance => XhrSandbox.isOpenedXhr || domUtils.isResponse(instance),
```
Comment (inikulin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Let's rename `isResponse` to `isFetchResponse`

File: src/client/utils/dom.js [hunk 7]
```diff
@@ -542,6 +542,30 @@ export function isSVGElementOrChild (el) {
     return !!closest(el, 'svg');
 }
 
+export function isHeaders (instance) {
```
Comment (inikulin) [inline src/client/utils/dom.js]: `isFetchHeaders`

File: src/client/utils/dom.js [hunk 8]
```diff
@@ -542,6 +542,30 @@ export function isSVGElementOrChild (el) {
...
+    if (nativeMethods.Headers && instance instanceof nativeMethods.Headers)
+        return true;
+
+    return instance && typeof instance === 'object' && instance.getAll !== void 0 &&
+           typeof instance.toString === 'function' && instance.toString() === '[object Headers]';
+}
+
+export function isRequest (instance) {
```
Comment (inikulin) [inline src/client/utils/dom.js]: `isFetchRequest`

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 9]
```diff
@@ -534,6 +534,24 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
                 }
             },
 
+            type: {
```
Comment (inikulin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Can we avoid using code instrumentation here?

**Comment by inikulin** [comment]
\r- Great job, but there are some minor issues

File: src/client/sandbox/code-instrumentation/properties/index.js (hunk 9 above)
Comment (miherlosev) [inline src/client/sandbox/code-instrumentation/properties/index.js]: No.
The `Response` constructor is not called during `fetch` request.
Example:
```javascript
    var nativeResponse = window.Response;

        window.Response = function () {
            alert('Response constructor was called');

            var response = null;

           if (!arguments.length)
                   response = new nativeResponse();
            else if(arguments.length === 1)
                   response = new nativeResponse(arguments[0]);
            else
                   response = new nativeResponse(arguments[0], arguments[1]);

            Object.defineProperty(response, 'type', {
                get: function () {
                    return 'overriden';
                }
            });
            return response;
        };

    fetch('http://localhost:3000')
            .then(function (response) {
                return response.text();
            })
            .then(function (body) {
                alert(body);
            })
            .catch(function (err) {
                alert(err);
            });
    </script>
```

**Comment by miherlosev** [comment]
FPR

File: src/client/sandbox/code-instrumentation/properties/index.js (hunk 9 above)
Comment (inikulin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: But can we instrument Promise returned by `fetch` and replace response property there?

File: src/client/sandbox/fetch.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            };
+
+            return originalThen.apply(this, args);
+        };
+    }
+
+    static _getResponseType (response) {
+        var parsedResponseUrl = parseProxyUrl(response.url);
+        var isSameOrigin      = sameOriginCheck(getDestLocation(),
+                                                parsedResponseUrl && parsedResponseUrl.destUrl,
```
Comment (inikulin) [inline src/client/sandbox/fetch.js]: Do not wrap args. Better use intermediate var for the condition.

File: src/client/sandbox/fetch.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        return response.status === 0 ? 'opaque' : 'cors';
+    }
+
+    attach (window) {
+        super.attach(window, window.document);
+
+        var sandbox = this;
+
+        if (window.fetch) {
+
```
Comment (inikulin) [inline src/client/sandbox/fetch.js]: excessive new line

**Comment by inikulin** [comment]
\r- Just some minor style issues left

**Comment by miherlosev** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by churkin** [comment]
lgtm

**Pull request merged**